src/customConfigTests/src/test/kotlin/org/vaccineimpact/orderlyweb/customConfigTests/GithubWebTests.kt

```diff
@@ -0,0 +1,42 @@
+package org.vaccineimpact.orderlyweb.customConfigTests
+
+import khttp.get
+import org.assertj.core.api.Assertions
+import org.junit.Test
+import org.vaccineimpact.orderlyweb.db.AppConfig
+
+
+class GithubWebTests : CustomConfigTests()
+{
+    val url: String = "http://localhost:${AppConfig()["app.port"]}/"
+    val loginUrl = "${url}login/"
+
+    @Test
+    fun `can redirect to Github when not logged in`()
+    {
+        startApp("auth.provider=github")
+
+        val response = get(url)
+
+        //khttp does redirection for you so we expect the response from github, not the original 302
+        Assertions.assertThat(response.statusCode).isEqualTo(200)
+        Assertions.assertThat(response.url).startsWith("https://github.com/login")
+
+        val match = "client_id=([^&]*)".toRegex().find(response.url)
+        Assertions.assertThat(match!!.groups[1]!!.value).isEqualTo(AppConfig()["auth.github_key"])
+
+    }
+
+    @Test
+    fun `login with invalid code and secret fails`()
+    {
+        startApp("auth.provider=github")
+
+        //This spoofs the callback by github after user has provided valid credentials
+        val fullUrl = "${loginUrl}?code=fake&secret=fake"
+
+        val response = get(fullUrl)
+
+        Assertions.assertThat(response.statusCode).isNotEqualTo(200)
```

Comment:
This comes back as a 500 from the pac4j built in authenticator rather than a 401 surprisingly. Perhaps because something's gone badly awry if it's getting something other than a legit request from the OAuth provider to this endpoint.

Reply:
Interesting, do you know what the error is? Which also reminds me, we should implement a "Something went wrong" page in the same model as our 404 page. Added a ticket: https://vimc.myjetbrains.com/youtrack/issue/mrc-216

Reply:
Here's the interesting bit of the call stack. I think what's happening is that it's trying to respond to the callback by going straight back to github to get the access token, but obviously not finding it in the rather unhappy response it gets from github. I'm a little surprised that it doesn't fail with an auth error before it attempts to make a request to github, since it should be able to tell immediately that the code and secret are duff. 
`[qtp806426226-13] ERROR org.vaccineimpact.orderlyweb.app_start.ErrorHandler - An unhandled exception occurred
[14:26:25][org.vaccineimpact.orderlyweb.customConfigTests.GithubWebTests.login with invalid code and secret fails]     com.github.scribejava.core.exceptions.OAuthException: Response body is incorrect. Can't extract a 'access_token=([^&]+)' from this: 'error=bad_verification_code&error_description=The+code+passed+is+incorrect+or+expired.&error_uri=https%3A%2F%2Fdeveloper.github.com%2Fapps%2Fmanaging-oauth-apps%2Ftroubleshooting-oauth-app-access-token-request-errors%2F%23bad-verification-code'
[14:26:25][org.vaccineimpact.orderlyweb.customConfigTests.GithubWebTests.login with invalid code and secret fails]     	at com.github.scribejava.core.extractors.OAuth2AccessTokenExtractor.extractParameter(OAuth2AccessTokenExtractor.java:70)
[14:26:25][org.vaccineimpact.orderlyweb.customConfigTests.GithubWebTests.login with invalid code and secret fails]     	at com.github.scribejava.core.extractors.OAuth2AccessTokenExtractor.extract(OAuth2AccessTokenExtractor.java:48)
[14:26:25][org.vaccineimpact.orderlyweb.customConfigTests.GithubWebTests.login with invalid code and secret fails]     	at com.github.scribejava.core.extractors.OAuth2AccessTokenExtractor.extract(OAuth2AccessTokenExtractor.java:16)
[14:26:25][org.vaccineimpact.orderlyweb.customConfigTests.GithubWebTests.login with invalid code and secret fails]     	at com.github.scribejava.core.oauth.OAuth20Service.sendAccessTokenRequestSync(OAuth20Service.java:44)
[14:26:25][org.vaccineimpact.orderlyweb.customConfigTests.GithubWebTests.login with invalid code and secret fails]     	at com.github.scribejava.core.oauth.OAuth20Service.getAccessToken(OAuth20Service.java:80)
[14:26:25][org.vaccineimpact.orderlyweb.customConfigTests.GithubWebTests.login with invalid code and secret fails]     	at com.github.scribejava.core.oauth.OAuth20Service.getAccessToken(OAuth20Service.java:73)
[14:26:25][org.vaccineimpact.orderlyweb.customConfigTests.GithubWebTests.login with invalid code and secret fails]     	at org.pac4j.oauth.credentials.authenticator.OAuth20Authenticator.retrieveAccessToken(OAuth20Authenticator.java:34)
[14:26:25][org.vaccineimpact.orderlyweb.customConfigTests.GithubWebTests.login with invalid code and secret fails]     	at org.pac4j.oauth.credentials.authenticator.OAuthAuthenticator.validate(OAuthAuthenticator.java:38)
[14:26:25][org.vaccineimpact.orderlyweb.customConfigTests.GithubWebTests.login with invalid code and secret fails]     	at org.vaccineimpact.orderlyweb.security.authentication.GithubOAuthAuthenticator.validate(GithubOAuthAuthenticator.kt:21)
[14:26:25][org.vaccineimpact.orderlyweb.customConfigTests.GithubWebTests.login with invalid code and secret fails]     	at org.vaccineimpact.orderlyweb.security.authentication.GithubOAuthAuthenticator.validate(GithubOAuthAuthenticator.kt:14)
[14:26:25][org.vaccineimpact.orderlyweb.customConfigTests.GithubWebTests.login with invalid code and secret fails]     	at org.pac4j.core.client.BaseClient.retrieveCredentials(BaseClient.java:71)
[14:26:25][org.vaccineimpact.orderlyweb.customConfigTests.GithubWebTests.login with invalid code and secret fails]     	at org.pac4j.core.client.IndirectClient.getCredentials(IndirectClient.java:140)
[14:26:25][org.vaccineimpact.orderlyweb.customConfigTests.GithubWebTests.login with invalid code and secret fails]     	at org.pac4j.core.engine.DefaultCallbackLogic.perform(DefaultCallbackLogic.java:89) `

Reply:
Well I guess a 500 is reasonable enough, as long as we handle it back to the user nicely.